plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+    (runReaderT (go t) $ Top^.depth)
+    -- initial table is empty
+    mempty
+ where
+   go :: Term tyname name uni a -> ReaderT Int (State (RhsTable tyname name uni a)) (Term tyname name uni a)
+   go = \case
+         -- this overrides the 'termSubterms' functionality only for the 'Let' constructor
+         LamAbs a n ty tBody -> LamAbs a n ty <$> local (+1) (go tBody)
+         TyAbs a n k tBody  -> TyAbs a n k <$> local (+1) (go tBody)
+
+         Let a r bs tIn -> do
+           curDepth <- ask
+           forM_ bs $ \ b -> do
+             b' <- bindingSubterms go b
+             lift . modify $ M.insert (b'^.principal) (a, r, b',curDepth)
+           go tIn
+         t' -> termSubterms go t'
+
+-- | Starts the 2nd pass from the 'Top' depth and the toplevel expression of the cleanedup term (devoid of any lets).
+-- See Note [Float second-pass]
+--
+p2Term :: forall name tyname uni a
+       . (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique, Monoid a)
+      => Term tyname name uni a --
+      -> FloatData
+      -> Term tyname name uni a
+p2Term pir fd =
+  -- the 2nd pass starts by trying to float any lets around the top-level expression (body)
+  -- For optimization reasons, we keep a state of remaining SCCs that we need to scan when we are generating let-groups in their correct order.
+  -- The initial state starts from the all sccs top.sorted; IMPORTANT: the invariant is that the sccs in the state should always remain top.sorted.
+  case runState (goFloat Top pirClean fd) topSortedSccs of
+    (res, []) -> res
+    (_, remState) -> error ("The final term is missing some lets because of a problem. The lets that could not be floated back to the final term were:" ++ show remState)
+
+ where
+  -- | Prior to starting the second pass, we clean the term from all its let-declarations and store them separately in a table.
+  -- The 2nd pass will later re-introduce these let-declarations, potentially placing them differently than before, thus essentially "floating the lets".
+  (pirClean :: Term tyname name uni a, rhsTable :: RhsTable tyname name uni a) = removeLets pir
+
+  -- 2nd-pass functions
+  ---------------------
+
+  -- | visit each term to apply the float transformation
+  -- OPTIMIZE: stop descending when state==mempty?
+  goTerm :: Int -- ^ current depth
+         -> FloatData -- ^ the lambdas we are searching for (to float some lets inside them)
+         -> Term tyname name uni a
+         -> State [S.Set PLC.Unique] (Term tyname name uni a)
+  goTerm curDepth searchTable = \case
+      -- we are only interested in lambdas/Lambdas
+      LamAbs a n ty tBody -> LamAbs a n ty <$> goAbs curDepth searchTable n tBody
+      TyAbs a n k tBody  -> TyAbs a n k <$> goAbs curDepth searchTable n tBody
+      -- descend otherwise to apply the transformations to subterms
+      t -> termSubterms (goTerm curDepth searchTable) t
```

Comment:
I think @michaelpj likes using `traverseOf` rather than feeding a function directly to the traversal. No strong feeling from my side, though.